### PR TITLE
Leader change epiphyte

### PR DESCRIPTION
**Pull request by vEpiphyte**
* Add a gated leader-change awareness that a cell can opt into.
* Add a stormdmon helper object to manage dmons so they can be started/stopped en masse
* Add a hook that a cortex can use for doing start/stop actions with

### CODE REVIEW

File: synapse/cortex.py [hunk 1]
```diff
@@ -912,8 +890,27 @@ async def finidmon():
...
+        # the leakership hook once at boot
+        self.leaderchangeaware = True
+        await self.iamLeaderHook()
+
+    async def iamLeaderHook(self):
+        if self.nexsroot.amLeader():
+            return await self.doLeaderStuff()
+        return await self.doNonLeaderStuff()
+
+    async def doLeaderStuff(self):
```
Comment (vEpiphyte) [inline synapse/cortex.py:906]: technically, the login in these could be rolled into `iamLeaderHook` but it does mean that overriding `iamLeaderHook` becomes more complicated

File: synapse/tests/test_cortex.py [hunk 2]
```diff
@@ -3799,15 +3794,16 @@ async def test_cortex_mirror(self):
...
-                # remove the mirrorness from the cortex
-                await core01.nexsroot.setLeader(None, None)
-                core01.mirror = False
-
-                async with await s_cortex.Cortex.anit(dirn=path00) as core00:
-                    self.len(1, await core01.nodes('[inet:ipv4=9.9.9.8]'))
-                    # add it back
-                    await core01.initCoreMirror(url)
-                    self.len(1, await core01.nodes('[inet:ipv4=9.9.9.9]'))
+                # Epiphyte: I don't think this is a valid test
```
Comment (vEpiphyte) [inline synapse/tests/test_cortex.py:3797]: FIXME: Restore this functionality

File: synapse/lib/storm.py [hunk 3]
```diff
@@ -783,6 +784,80 @@
...
+    async def stop(self):
+        '''
+        Stop all the dmons.
+        '''
+        if not self.enabled:
+            return
+        await self._stopAllDmons()
+        await asyncio.sleep(0)
+
+    # TODO write enable/disable APIS.
```
Comment (vEpiphyte) [inline synapse/lib/storm.py:853]: This is left as a TODO for another story

File: synapse/lib/storm.py [hunk 4]
```diff
@@ -783,6 +784,80 @@
...
+        if dmon:
+            await dmon.fini()
+
+    async def start(self):
+        '''
+        Start all the dmons.
+        '''
+        if self.enabled:
+            return
+        for dmon in list(self.dmons.values()):
```
Comment (vEpiphyte) [inline synapse/lib/storm.py:840]: We have a turn-on/turn-off/turn-on problem here where the underlying dmon objects will ahve been fini'd

File: synapse/cortex.py [hunk 5]
```diff
@@ -912,8 +890,27 @@ async def finidmon():
 
         await self.postNexsAnit()
 
-        if mirror is not None:
-            await self.initCoreMirror(mirror)
+        if self.mirror is not None:
+            await self._initCoreMirror(self.mirror)
+
+        # Enable leadership change awareness and fire
+        # the leakership hook once at boot
```
Comment (jnwatson) [inline synapse/cortex.py:897]: leakership (sp)

File: synapse/lib/cell.py [hunk 6]
```diff
@@ -575,6 +577,13 @@ async def _initNexsRoot(self):
         nexsroot.onfini(self)
         return nexsroot
 
+    async def _leaderHookGate(self):
```
Comment (jnwatson) [inline synapse/lib/cell.py:580]: I still don't understand why you have two methods here. Just make 1. You're already taking the method call hit.

**Pull request merged**